lib/mongo/socket.rb

```diff
@@ -169,7 +169,22 @@ def eof?
     private
 
     def read_from_socket(length)
-      @socket.read(length) || String.new
+      # Block on data to read for timeout seconds
+      # using the suggested implementation of http://www.ruby-doc.org/core-2.1.3/Kernel.html#method-i-select
+      # to work with SSL connections and some OSs like Ubuntu, CentOS and Red Hat
+      time_left = timeout.to_f
+      data = ""
+      begin
+        while data.length < length
+          data << @socket.read_nonblock(length - data.length)
+        end
+      rescue IO::WaitReadable
+        raise Error::OperationFailure.new("Took more than #{timeout} seconds to receive data.") if (time_left -= 0.1) <= 0
```

Comment:
Actually I think here we would want to raise a `TimeoutError` instead of an `OperationFailure`, to be consistent with the previous behaviour (which raised a `TimeoutError` but wasn't doing it within the actual timeout set on the initial socket, at least on OSX for me locally)